### ISSUE TITLE
[Specification] OpImageFetch with ConstOffset

### ISSUE THREAD

**Issue by Retrofractal** [KhronosGroup/Vulkan-Docs#1416]
`ConstOffset` operation does not appear to be defined for `OpImageFetch`. The offsets are applied in "(s,t,r,q,a) to (u,v,w,a) Transformation", which doesn't cover texel fetches. Interestingly, it also doesn't appear to be adequately described for `texelFetchOffset` in the OpenGL specification, for the same reason.

**Comment by gnl21** [comment]
I agree that this isn't adequately covered in the spec but it is supposed to work. It appears to be tested in the CTS.

The spec doesn't currently say which image operation steps are performed by `OpImageFetch`, but I assume that its coordinates are supposed to be interpreted as either unnormalized or integer coordinates, and the offset should be applied at that stage. (I don't think it actually makes a difference which stage it happens in).

**Comment by Retrofractal** [comment]
Texel fetches don't make use of the sampler state so only integer coordinates are permitted. Nevertheless, as you imply, the issue here is that the offsets are only applied to normalized coordinates, which isn't much good! Section 16.10 indicates which of the instructions apply the normalized to unnormalized and the unnormalized to integer texel coordinate transformations:
> (s,t,r,q,a) to (u,v,w,a) Transformation, Wrapping, and (u,v,w,a) to (i,j,k,l,n) Transformation And Array Layer Selection: Performed by all OpImageSample, OpImageSparseSample, and OpImage*Gather instructions.

**Comment by TomOlson** [comment]
Agreed the spec needs clarification, so adding it to our internal queue. Not considered urgent since implementations are at least consistent, but we will fix the spec description as soon as we can.

**Comment by gnl21** [comment]
> Texel fetches don't make use of the sampler state so only integer coordinates are permitted. [...] Section 16.10 indicates ...

Yes, I had misread that section and thought that none of the bullets applied to `OpImageFetch`. You're right that the expectation seems to be that the coordinates are treated as integer coordinates (and that makes most sense, as you say, since they don't use sampler state and the coordinates are formatted as integers in the shader). That means that the SPIR-V spec has the same issue, since it talks about applying the offsets to unnormalized coordinates as well.

**Comment by Retrofractal** [comment]
OK, thanks, for what it's worth, I should think the best solution is to move the offset deltas from section 16.5.8 into section 16.7 under a new sub-heading that makes it clear that the offset operation applies to all integer coordinates (not just operations with integer inputs) and prior to texel input validation. This is assuming it's understood that unnormalized coordinates always undergo a transformation to integer coordinates as described in section 16.6 but omitted from the opening paragraph of section 16.1.1.

> That means that the SPIR-V spec has the same issue, since it talks about applying the offsets to unnormalized coordinates as well.

I can see that the optional image operands descriptions claim the offsets are added to the `(u, v, w)` coordinates, which isn't true for normalized coordinates, if we're assuming that refers to the operation input rather than the coordinates defined in section 16.1.1. in the Vulkan spec. Were you referring to that or some other section?

**Comment by gnl21** [comment]
> I should think the best solution is to move the offset deltas from section 16.5.8 into section 16.7 under a new sub-heading that makes it clear that the offset operation applies to all integer coordinates

Yes, this is very similar to what I was thinking.

> This is assuming it's understood that unnormalized coordinates always undergo a transformation to integer coordinates as described in section 16.6 but omitted from the opening paragraph of section 16.1.1.

I generally agree with that interpretation, although it is complicated by the fact that the translation from unnormalized to integer coordinates may produce multiple outputs (for filtered lookups, etc.), so that might make implementing it in the spec like that less nice.

> Were you referring to that or some other section?

I was referring to that section, but the issue I was talking about was with texelFetch, since it bypasses the unnormalized coordinates stage. (I interpreted normalized coordinates as still going through the unnormalized stage and being offset there). If you follow the interpretation from your suggested solution above then this just needs changing to say "integer coordinates (i, j, k)", but I'll fix that spec once the Vulkan change is done.

**Comment by Retrofractal** [comment]
Ah OK, so you were thinking in terms of the Vulkan spec coordinate symbols instead of the SPIR-V `Coordinate` parameter. Makes sense, thanks very much for looking into this 👍
